### ISSUE TITLE
CuDNN v4.0 BatchNormalization output wrong data in [Test phrase]

### ISSUE THREAD

**Issue by pl8787** [NVIDIA/caffe#94]
CuDNN v4.0 BatchNormalization output wrong data in [Test phrase].
It works ok in [Train phrase].

That means

```C++
CUDNN_CHECK(cudnnBatchNormalizationForwardInference(
      Caffe::cudnn_handle(),
      mode_,
      cudnn::dataType<Dtype>::one,
      cudnn::dataType<Dtype>::zero,
      bottom_desc_,
      bottom_data,
      bottom_desc_,
      top_data,
      scale_bias_mean_var_desc_,
      scale_data,
      bias_data,
      this->blobs_[3]->gpu_data(),  // mean
      this->blobs_[4]->gpu_data(),  // variance
      epsilon_));
```

return wrong values.

We use

```C++
CUDNN_CHECK(cudnnBatchNormalizationForwardTraining(
      Caffe::cudnn_handle(),
      mode_,
      cudnn::dataType<Dtype>::one,
      cudnn::dataType<Dtype>::zero,
      bottom_desc_,
      bottom_data,
      bottom_desc_,
      top_data,
      scale_bias_mean_var_desc_,
      scale_data,
      bias_data,
      1,
      this->blobs_[3]->mutable_gpu_data(),  // mean
      this->blobs_[4]->mutable_gpu_data(),  // variance
      epsilon_,
      save_mean,
      save_inv_var));
```

instead, that work fine.

**Comment by andrei-pokrovsky** [comment]
Hi! Thank you for reporting the issue. So generally the output from these two functions is going to be different, the first one will use the values for mean and variance accumulated during training as constants, as defined in the paper, Algorithm 2, step 11. By second function working fine, do you mean that it produces the same results as Caffe's implementation? If using the second function works fine that would seem to indicate that Caffe's implementation is different from what the original BN paper is doing because the mean and variance in the ForwardTraining version are computed from the batch data only (no moving average). Are you getting worse classification accuracy using the ForwardInference version?

**Comment by zhangfd** [comment]
Thanks for the reply! We got very bad accuracy using ForwardInference. We found the  value became  very  small  after  the function  in  TEST phase. So we tried ForwardTraining just  to debug since there seem to be no significant distinguish at  first  few  iterations(not sure whether it's right). We tried  to  stop moving average by setting the  factor to 1 but we  are  not sure that's correct. So is there any problems of the ForwardInference function or we use it incorrectly or weforgot  something when  merging the code？Thank you!

**Comment by andrei-pokrovsky** [comment]
Have you tried running the same network and dataset with engine: CAFFE parameter for the BN layer? Do you get higher accuracy? Also I'm assuming you get higher accuracy with Caffe's BN than without BN at all? We reran some simple convergence tests here on Cifar10 with CAFFE and CUDNN layers and get similar results with both CAFFE and CUDNN. We'll keep investigating.

**Comment by zhangfd** [comment]
@andrei-pokrovsky I haven't tried CAFFE's BN yet. We get much higher valid accuracy without BN at all (85%vs80% on our 18000-image valid set), but get similar train error. Is it because of the 'moving_average_fraction'? We used the default value 0.999. It seems that it needs all most the most recent 10,000 batches to estimate over 99.99% of the final moving mean and variance. Thanks!

**Comment by andrei-pokrovsky** [comment]
Are you using the layer in-place? (meaning bottom and top layers for BN have the same name)
That doesn't work correctly right now in Caffe for some reason even though cuDNN's implementation supports in-place computation.

Caffe's default value of 0.999 should fine. Note that cudnn uses a 1-moving_average_fraction because of the difference in the API. cuDNN is doing avg = avg*(1-factor) + newVal*factor whereas Caffe defines the moving avg as inverted.

**Comment by andrei-pokrovsky** [comment]
If you could send me an email directly at a1pokrovsky@nvidia.com (without the 1) we could try to figure out how to create a repro. If you could share your model/dataset or create a repro on a public dataset it would definitely help. Something that would definitely help is if you could try your model with stock Caffe BN and see if it works.

**Comment by zhangfd** [comment]
@andrei-pokrovsky Thank you for the reply! I sent you an email.
